src/dataset.ts

```diff
@@ -341,6 +347,41 @@ export function checkStructure(s: JsObject): string {
         }
     }
 
+    if ('shape' in s) {
+        const array = s['shape'];
+        if (!Array.isArray(array)) {
+            return `"shape" must be an array`;
+        }
+        if (s.size > 0 && array.length !== s.size) {
+            return `wrong size for "shape", expected ${s.size}, got ${array.length}`;
+        }
+        for (var a of array) {
```

Comment:
Don't use `var`, it has weird scoping rules

```suggestion
        for (const a of array) {
```